### PRE-LECTURE HW

### Q1

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

this is the code chatgpt gave me to check dataset has missing values

In [1]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Count the number of missing values in each column
missing_values = df.isna().sum()
print(missing_values)


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


### Q2

chatgpt gave me this code: 

In [2]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = df.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 391
Number of columns: 11


Regarding meaning of "observations" and "variables"

Observations: These are the individual units or cases in a dataset, typically represented by rows in a table. Each observation corresponds to a single data point or entity. For example, in a dataset about villagers, each villager (with their specific attributes) would be an observation.

Variables: These are the characteristics or attributes that are measured or recorded for each observation, typically represented by columns in a dataset. For instance, a dataset about villagers might include variables like age, occupation, and species.

### Q3
chat gpt:

In [ ]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# 1. Summary Statistics
summary_stats = df.describe()

# 2. Basic Info
basic_info = df.info()

# 3. Unique Values per Column
unique_values = df.nunique()

# 4. Value Counts for a specific column (e.g., 'species')
value_counts = df['species'].value_counts()

# Display results
print("Summary Statistics:\n", summary_stats)
print("\nBasic Info:\n", basic_info)
print("\nUnique Values per Column:\n", unique_values)
print("\nValue Counts for 'species' Column:\n", value_counts)


### Q4

1. Number of Columns Analyzed
df.shape: This function gives the total number of rows (observations) and columns (variables) in the dataset. It includes all columns, regardless of their data type (numeric, categorical, etc.).

df.describe(): This function by default only analyzes numeric columns (e.g., integers and floats). If the dataset contains other types of columns, such as strings (categorical data), df.describe() will ignore them unless you use the include='all' parameter, which includes all columns (numeric, categorical, etc.).

2. Values Reported in the "Count" Column

df.describe() reports a "count" for each numeric column. This "count" represents the number of non-missing (non-NaN) values for each column. So, if some columns have missing data (NaN values), the count for those columns in df.describe() will be lower than the total number of rows in df.shape.


### Q5

The key difference between an "attribute" and a "method" in Python (especially when using libraries like pandas) lies in how they function and are used:

### 1. Attribute
An attribute is a property or characteristic associated with an object. It stores data or values related to that object.
It does not require parentheses (()).
Accessing an attribute simply returns the stored value; it does not perform any computation or operation.

df.shape
df.shape is an attribute that returns the dimensions of the DataFrame (number of rows and columns). It directly gives a tuple like (rows, columns) without performing any action.

### 2. Method
A method is a function that is associated with an object and typically performs an operation or computation on that object.
It requires parentheses (()), even if no parameters are passed, to indicate that the method is being called to execute some action.
Methods can optionally accept arguments that modify their behavior.


df.describe()
df.describe() is a method that calculates and returns statistical summaries for the numeric columns in the DataFrame. It performs an action on the data and returns a result, so the parentheses are needed to "call" the method.

This is the chatgpt answer however my understanding is:
Attributes like df.shape() store data about the object, and no parentheses are used.
Methods like df.describe() perform operations or calculations, and parentheses are required to call them.

### POST-LECTURE HW

### Q6

Count: Represents the number of non-missing values in a column.
Mean: The average value of the column.
Std (Standard Deviation): A measure of the dispersion or spread of the data around the mean.
Min: The minimum value in the column (smallest non-missing value).
25% (1st Quartile): The first quartile, which means that 25% of the values in the column are less than or equal to this value.
50% (Median or 2nd Quartile): The median value, or the middle value when the data are sorted in order.
75% (3rd Quartile): The third quartile, meaning 75% of the values are less than or equal to this value.
Max: The maximum value in the column (largest non-missing value).

### Q7

1. We want to remove rows from a dataset with multiple columns that have missing values in any of these columns because the analysis of those rows would be erroneous.The reason this method is preferred is that it only removes rows where one or more cells have missing data, while maintaining the structure of all columns.

2. large dataset where some columns are considered irrelevant or needless for analysis because they have a high number of missing values.

    When deciding which columns to remove because they are either unnecessary or have an excessive number of missing values, this approach is recommended.

3. It may be important to apply del df['col'] before df.dropna() in order to avoid needlessly losing important data in other columns by eliminating superfluous or significantly missing columns first.

4. approach:

   Remove the Cabin Column Entirely: Since the Cabin column has a significant number of missing values and is generally less informative for most analyses (e.g., predicting survival), it makes sense to delete this column altogether.

   Handle the Age Column Separately





i asked chatgpt for the code:

In [5]:

# Cleaning the data
threshold = 0.5 * len(df)
df_cleaned = df.dropna(thresh=threshold, axis=1)
df_cleaned = df_cleaned.dropna()

# Show cleaned data
print("\nAfter Data Cleaning:")
print(df_cleaned.head())
print("\nNumber of rows after cleaning:", df_cleaned.shape[0])
print("Number of columns after cleaning:", df_cleaned.shape[1])



After Data Cleaning:
   row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  ht

### Q8

In my own Words:

what "df.groupby("col1")["col2"].describe()" does is generate and show "count, mean, ..., max" of col2, for each unique value of col1.

Summary of Imaginary Dataset and Demonstration

We created an imaginary dataset to analyze student GPAs from four different universities: University of Toronto (UofT), University of Waterloo (UWaterloo), York University (York), and Western University (Western). Each university had 20 students, making a total of 80 students in the dataset.

Dataset Structure:

University: Represents the name of the university (UofT, UWaterloo, York, Western).
GPA: Represents the GPA of each student, randomly generated based on a normal distribution with different means and standard deviations for each university.
Objective:

To demonstrate the use of df.groupby("col1")["col2"].describe() to analyze the distribution of GPAs (GPA) for students from each university (University).
Here is the ChatGPT code bellow:



In [6]:
import pandas as pd
import numpy as np

# Set a random seed for reproducibility
np.random.seed(42)

# Create an imaginary dataset with 100 students (20 from each university)
data = {
    'University': (['UofT'] * 20) + (['UWaterloo'] * 20) + (['York'] * 20) + (['Western'] * 20),
    'GPA': np.concatenate([
        np.random.normal(3.5, 0.3, 20),  # GPAs for UofT students
        np.random.normal(3.2, 0.4, 20),  # GPAs for UWaterloo students
        np.random.normal(2.8, 0.5, 20),  # GPAs for York students
        np.random.normal(3.0, 0.6, 20)   # GPAs for Western students
    ])
}

# Create a DataFrame
df_students = pd.DataFrame(data)

# Group by 'University' and describe the 'GPA' column
gpa_description_by_university = df_students.groupby("University")["GPA"].describe()

# Display the result
print(gpa_description_by_university)

            count      mean       std       min       25%       50%       75%  \
University                                                                      
UWaterloo    20.0  3.093610  0.387215  2.416132  2.767587  3.096506  3.279945   
UofT         20.0  3.448610  0.288009  2.926016  3.305384  3.429756  3.652453   
Western      20.0  2.981162  0.667253  1.428153  2.687604  3.015366  3.488922   
York         20.0  2.786655  0.410424  1.918480  2.542645  2.813930  3.121687   

                 max  
University            
UWaterloo   3.940911  
UofT        3.973764  
Western     3.938786  
York        3.328561  


### Q8 - 2

Difference Between df.describe() and df.groupby("col1")["col2"].describe() Counts:

df.describe() Count:

Shows the number of non-missing (non-NaN) values for each column in the entire dataset.
If a column has missing data, its count will be less than the total number of rows. This tells you how much data is available for each column overall.
df.groupby("col1")["col2"].describe() Count:

Shows the number of non-missing values for col2 within each group defined by the unique values of col1.
It tells you how much data is available for col2 in each category of col1 (e.g., grouping by passenger class and counting ages in the Titanic dataset).
Key Difference:

df.describe() provides a summary of missing data for each column in the entire dataset.
df.groupby("col1")["col2"].describe() provides a summary of missing data for col2 within each group of col1.
Why This Matters:

df.describe() helps you see overall data completeness for each column.
df.groupby("col1")["col2"].describe() helps you understand data completeness and distribution within specific groups, which is useful for more detailed analysis.


# Q8 -3

## Q8 A

In [17]:
# Omitted: import pandas as pd

# Create a DataFrame without importing pandas
data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
df = pd.DataFrame(data)  # This will raise an error

In [18]:
import pandas as pd

# Now this works correctly
data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
df = pd.DataFrame(data)
print(df)

   A  B
0  1  4
1  2  5
2  3  6


## Q8 B

In [19]:
import pandas as pd

# Mistyped filename in the URL
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanics.csv'  # Incorrect filename 'titanics.csv'
titanic_data = pd.read_csv(url)  # This will raise an error


HTTPError: HTTP Error 404: Not Found

In [ ]:
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'  # Correct filename 'titanic.csv'
titanic_data = pd.read_csv(url)

## Q8 C

In [ ]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

In [20]:
# Incorrectly using the DataFrame variable
# Intentional typo in the variable name (should be 'titanic_data')
Titanic_Data.groupby("Pclass")["Age"].describe()  # This will raise an error

NameError: name 'Titanic_Data' is not defined

In [21]:
# Correctly using the DataFrame variable
titanic_data.groupby("Pclass")["Age"].describe()  # Correct usage

count       mean        std   min   25%   50%   75%   max
Pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

## Q8 D

In [22]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# Forgetting the closing parenthesis in the read_csv function
titanic_data = pd.read_csv(url  # Missing closing parenthesis

SyntaxError: incomplete input (4063547765.py, line 7)

In [23]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# Correcting by adding the closing parenthesis
titanic_data = pd.read_csv(url)  # Now this works correctly

## Q8 E

In [24]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Mistyping 'groupby' as 'group_by'
titanic_data.group_by("Pclass")["Age"].describe()  # This will raise an error


AttributeError: 'DataFrame' object has no attribute 'group_by'

In [25]:
# Correcting the function name 'groupby'
titanic_data.groupby("Pclass")["Age"].describe()  # Now this works correctly

count       mean        std   min   25%   50%   75%   max
Pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [26]:
# Mistyping 'describe' as 'describle'
titanic_data.groupby("Pclass")["Age"].describle()  # This will raise an error

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

In [27]:
# Correcting the function name 'describe'
titanic_data.groupby("Pclass")["Age"].describe()  # Now this works correctly

count       mean        std   min   25%   50%   75%   max
Pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

## Q8 F

In [28]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Mistyping the column name 'Sex' as 'sex' in groupby
titanic_data.groupby("sex")["Age"].describe()  # This will raise an error

KeyError: 'sex'

In [29]:
# Correct column names for groupby and selection
titanic_data.groupby("Sex")["Age"].describe()  # Correct groupby column 'Sex'

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [30]:
# Mistyping the column name 'age' as 'Age' in the column selection
titanic_data.groupby("sex")["age"].describe()  # This will raise an error

KeyError: 'sex'

In [31]:
# Correct column names for groupby and selection
titanic_data.groupby("Sex")["Age"].describe()  # Correct column selection 'Age'

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

## Q8 G

In [32]:
import pandas as pd

# Correct URL to load the Titanic dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Missing quotes around 'sex' in groupby
titanic_data.groupby(sex)["Age"].describe()  # This will raise an error


NameError: name 'sex' is not defined

In [33]:
# Correct column name usage with quotes
titanic_data.groupby("Sex")["Age"].describe()  # Correct usage with quotes

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [34]:
# Missing quotes around 'Age' in column selection
titanic_data.groupby("sex")[Age].describe()  # This will raise an error

KeyError: 'sex'

In [35]:
# Correct column name usage with quotes
titanic_data.groupby("Sex")["Age"].describe()  # Correct usage with quotes

count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

Troubleshooting Code: ChatGPT vs. Google
Alright, here’s the deal. When I’m stuck with coding issues, I’d rather use ChatGPT over Google. Why? Because with Google, I have to dig through a bunch of websites, read endless threads, and sometimes still not find exactly what I need. It’s like a scavenger hunt.
With ChatGPT, it’s more like having a friend who knows their stuff and walks me through the problem step-by-step. It’s way more direct and less frustrating. Plus, I can ask follow-up questions without jumping from page to page. So yeah, ChatGPT feels more like a human tutor, making troubleshooting less of a headache.


## Q9

kind off

Here’s a summary of our chat: part 1

You worked with a dataset from a CSV file using pandas, starting with checking for missing values (df.isna().sum()).
I explained how to determine the number of rows and columns in the dataset using df.shape.
We discussed the meaning of terms like observations (rows) and variables (columns).
I provided ways to generate summaries of the columns in the dataset, using functions like .describe(), .info(), and .value_counts().
We explored the discrepancy between df.shape (total columns/rows) and df.describe() (which only summarizes numeric columns by default and reports counts of non-missing values).
I explained the difference between an attribute (e.g., df.shape) and a method (e.g., df.describe()), emphasizing how attributes store data, while methods perform operations.
Finally, we reviewed the statistical terms provided by df.describe(), including count, mean, std, min, 25%, 50%, 75%, and max.

Here’s a summary of our conversation: part 2

Loading Data: We started by loading the Titanic dataset from an online source using pd.read_csv().
Groupby and Describe: You asked about df.groupby("col1")["col2"].describe(), which groups the DataFrame by col1 and provides descriptive statistics for col2 within each group.
Example with Titanic Dataset: We grouped the Titanic data by Pclass (passenger class) and described the Age column. This provided summary statistics (e.g., count, mean, min, max) for the ages in each passenger class.
Simplified Version: A simplified version of the code was provided, focusing on how to load the data, group by Pclass, and get descriptive statistics for Age.
Summary of Code: We wrapped up with a concise explanation of the code, explaining how it helps to analyze the distribution of age by passenger class on the Titanic.


Here is the link to my ChatGPT intraction of my first one:
https://chatgpt.com/share/87d16250-6259-4e38-8cb9-d741351bd03b

Here is the link to my ChatGPT intraction of my second one:
https://chatgpt.com/share/8e09fdc2-70cb-4ad7-a8b2-6c2c523739fe

lastly, i used chatgpt alongside my friend, hence why someone chatgpt responses are missing.


### Chatbot Summary: Data Cleaning and Error Handling with the Titanic Dataset

#### Objective:
To clean the Titanic dataset by handling missing values effectively and then intentionally introduce and fix common coding errors to improve data analysis and debugging skills.

#### Key Steps Taken:

1. **Understanding the Data**:
   - Reviewed the Titanic dataset's structure, identifying that the `Cabin` and `Age` columns had a significant number of missing values, while most other columns had very few or none.

2. **Decided on the Data Cleaning Approach**:
   - **Remove the `Cabin` column**: Due to a high number of missing values (687 out of 891) and its low relevance to most analyses.
   - **Impute missing values in the `Age` column**: Fill missing `Age` values with the median to maintain data integrity without introducing bias.
   - **Drop remaining rows with missing values**: After handling `Cabin` and `Age`, drop any rows with missing values in other columns to ensure a clean dataset.

3. **Implemented the Data Cleaning Steps**:
   - Removed the `Cabin` column using `del df['Cabin']`.
   - Imputed missing `Age` values using the median with `df['Age'].fillna(df['Age'].median(), inplace=True)`.
   - Dropped any remaining rows with missing values using `df.dropna(inplace=True)`.

4. **Created a Simplified Before and After Report**:
   - Introduced a concise reporting function to provide a clear summary of the dataset's state before and after cleaning, including the number of rows, columns, and missing values per column.

#### Results:
- The dataset was successfully cleaned, with all missing values addressed. The `Cabin` column was removed, `Age` values were imputed, and remaining rows with missing values were dropped, resulting in a dataset ready for further analysis.

#### Next Steps:
- Proceed with more advanced data analysis, such as feature engineering, data visualization, or building predictive models using the cleaned Titanic dataset.

---

### Error Handling and Debugging Exercises

After cleaning the dataset, we explored various intentional errors and their fixes to enhance our understanding of common coding issues and debugging techniques:

1. **Forget to Import a Library**:
   - **Error Created**: Removed the `import pandas as pd` line, which caused a `NameError: name 'pd' is not defined`.
   - **Solution**: Re-added the `import pandas as pd` line to fix the error.

2. **Incorrect Filename in URL**:
   - **Error Created**: Changed the filename in the URL from `titanic.csv` to `titanics.csv`, resulting in a `FileNotFoundError`.
   - **Solution**: Corrected the filename in the URL back to `titanic.csv`.

3. **Typo in URL Domain or Path**:
   - **Error Created**: Introduced typos in different parts of the URL (e.g., domain name, path), resulting in various errors like `URLError` and `HTTPError 404: Not Found`.
   - **Solution**: Corrected the typos in the URL.

4. **Using a Variable Before Assignment**:
   - **Error Created**: Misnamed the DataFrame variable, e.g., `Titanic_Data` instead of `titanic_data`, causing a `NameError`.
   - **Solution**: Corrected the variable name to `titanic_data`.

5. **Missing Parenthesis in Function Call**:
   - **Error Created**: Removed a closing parenthesis in a function call like `pd.read_csv(url`, resulting in a `SyntaxError: unexpected EOF while parsing`.
   - **Solution**: Added the missing parenthesis to complete the function call.

6. **Mistyped Function Names**:
   - **Error Created**: Misspelled function names like `group_by` instead of `groupby` and `describle` instead of `describe`, causing `AttributeError`.
   - **Solution**: Corrected the function names to `groupby` and `describe`.

7. **Using a Column Name Not in the Data**:
   - **Error Created**: Used incorrect column names in `groupby` and column selection, e.g., `"Sex"` instead of `"sex"` and `"age"` instead of `"Age"`, resulting in `KeyError`.
   - **Solution**: Corrected the column names to match exactly those in the DataFrame.

8. **Forgetting to Put Column Names in Quotes**:
   - **Error Created**: Omitted quotes around column names, such as `groupby(sex)` instead of `groupby("sex")`, causing a `NameError`.
   - **Solution**: Added quotes around column names to ensure they are treated as strings.

### Conclusion:
This exercise demonstrated how to clean a dataset effectively and handle missing values while providing practical experience in identifying, creating, and fixing common programming errors. The process strengthened understanding of debugging techniques and helped develop best practices for data analysis using pandas.
